girder/models/file.py

```diff
@@ -94,10 +95,19 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type extraParameters: str or None
         """
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
+            fileDownload = self.getAssetstoreAdapter(file).downloadFile(
                 file, offset=offset, headers=headers, endByte=endByte,
                 contentDisposition=contentDisposition,
                 extraParameters=extraParameters)
+            if inspect.isgeneratorfunction(fileDownload):
+                def downloadGenerator():
+                    for data in fileDownload():
+                        yield data
+                    if endByte is None or endByte >= file['size']:
+                        events.trigger('download.complete', info={'id': file['_id']})
+                return downloadGenerator
+            else:
+                return fileDownload
```

Comment:
Trigger the download.complete in both occurrences of the HTTPRedirect like this? 
```python
if file.get('assetstoreId'):
    try:
        fileDownload = self.getAssetstoreAdapter(file).downloadFile(
            file, offset=offset, headers=headers, endByte=endByte,
            contentDisposition=contentDisposition,
            extraParameters=extraParameters)
        if inspect.isgeneratorfunction(fileDownload) or inspect.isgenerator(fileDownload):
            def downloadGenerator():
                for data in fileDownload():
                    yield data
                if endByte is None or endByte >= file['size']:
                    events.trigger('download.complete', info={'file': file})
            return downloadGenerator
        else:
            return fileDownload
    except cherrypy.HTTPRedirect as e:
        events.trigger('download.complete', info={'file': file, 'redirect': True})
        raise e
elif file.get('linkUrl'):
    if headers:
        events.trigger('download.complete', info={'file': file, 'redirect': True})
        raise cherrypy.HTTPRedirect(file['linkUrl'])
    else:
        endByte = endByte or len(file['linkUrl'])

        def stream():
            yield file['linkUrl'][offset:endByte]
            if endByte >= len(file['linkUrl']):
                events.trigger('download.complete', info={'file': file})
        return stream
else:  # pragma: no cover
    raise Exception('File has no known download mechanism.')
```